File: ui/src/lib/api/participants.ts
/**
 * Typed wrapper for the unified participant directory (AUDIT.md §3.1).
 *
 *   - listParticipants(teamId) -> GET /v1/participants
 *
 * `GET /v1/participants` is the single, non-admin roster of EVERYONE on a team:
 * both humans (`kind:"human"`) and agents (`kind:"agent"`). It supersedes the
 * old split between `GET /v1/agents` (agents-only, presence) and the admin-only
 * `GET /v1/teams/{id}/members` (humans, no name, 403 to non-admins).
 *
 * The definitive human-vs-agent signal is `kind`, derived AUTHORITATIVELY by
 * the server from `agents.agent_type`. The UI keys on `kind` and NEVER guesses
 * by matching aliases against another roster.
 *
 * This is the source for: the team roster, the chat recipient picker, and the
 * issue assignee picker. `alias` is the team-unique stable selector used as
 * both a chat `to_aliases` target and an issue `assignee_id`.
 *
 * Goes through the shared `authedRequest` helper (Bearer JWT + the
 * `X-AWEB-Team-Id` header), so the team id is passed through explicitly.
 */

import { authedRequest } from "./http";

/** Authoritative participant kind, derived by the server from `agent_type`. */
export type ParticipantKind = "human" | "agent";

/**
 * A single team participant (human or agent). `kind` is authoritative; the UI
 * keys on it and never guesses. Presence applies to any participant that
 * heartbeats: a signed-in human (via POST /v1/presence/heartbeat) and an agent
 * (via its own heartbeat) both carry live `online`/`status`/`last_seen`.
 */
export interface Participant {
  /** AUTHORITATIVE. "human" iff agent_type === 'human', else "agent". */
  kind: ParticipantKind;
  /** Team-unique stable selector. Used as chat `to_aliases` + issue `assignee_id`. */
  alias: string;
  /** human_name for humans; alias for agents when blank. */
  display_name: string;
  /** agents.agent_id (UUID) when present. */
  agent_id: string | null;
  did_key: string | null;
  did_aw: string | null;
  /** "team_id/alias". */
  address: string | null;
  role: string | null;
  /** Raw passthrough of agents.agent_type ("human" | "agent" | ...). */
  agent_type: string;
  /** True for any participant (human or agent) with live presence. */
  online: boolean;
  /** "active" when a fresh presence record exists, else "offline". */
  status: string;
  /** ISO-8601 last-seen, or null when no presence record exists. */
  last_seen: string | null;
}

export interface ListParticipantsResponse {
  team_id: string;
  participants: Participant[];
}

/**
 * List ALL participants in a team — humans and agents — with an authoritative
 * `kind`. Non-admin: available to any team-scoped caller. Returns [] when no
 * team is selected.
 */
export async function listParticipants(
  teamId: string | null,
): Promise<Participant[]> {
  if (!teamId) return [];
  const res = await authedRequest<ListParticipantsResponse>("/v1/participants", {
    teamId,
  });
  return res.participants ?? [];
}

File: server/src/aweb/routes/participants.py
"""Unified participant directory: ``GET /v1/participants``.

The ``agents`` table is the participant/identity directory, not an agent-only
table. A human becomes a first-class chat/comment/assignee participant by having
a row here with ``agent_type='human'``. This endpoint exposes BOTH humans and
agents to ANY team member (it is NOT admin-gated), each with an authoritative
``kind`` ("human" | "agent") derived from ``agents.agent_type``.

The UI uses this endpoint for the team roster, the chat recipient picker, and
the issue assignee picker. ``GET /v1/agents`` stays unchanged (agents-only, with
presence) for back-compat.
"""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel

from aweb.deps import get_db, get_redis
from aweb.team_auth_deps import TeamIdentity, get_team_identity

from ..presence import list_agent_presences_by_workspace_ids

router = APIRouter(prefix="/v1/participants", tags=["participants"])


class ParticipantView(BaseModel):
    kind: str  # "human" | "agent" — AUTHORITATIVE; UI keys on this, never guesses
    alias: str  # team-unique selector (chat to_aliases + issue assignee_id)
    display_name: str  # human_name for humans; alias for agents (or blank human_name)
    agent_id: Optional[str] = None
    did_key: Optional[str] = None
    did_aw: Optional[str] = None
    address: Optional[str] = None
    role: Optional[str] = None
    agent_type: str = "agent"  # raw passthrough of agents.agent_type
    online: bool = False
    status: str = "offline"
    last_seen: Optional[str] = None


class ListParticipantsResponse(BaseModel):
    team_id: str
    participants: list[ParticipantView]


def _participant_kind(agent_type: Optional[str]) -> str:
    return "human" if (agent_type or "agent") == "human" else "agent"


@router.get("", response_model=ListParticipantsResponse)
async def list_participants(
    request: Request,
    db=Depends(get_db),
    redis=Depends(get_redis),
    identity: TeamIdentity = Depends(get_team_identity),
) -> ListParticipantsResponse:
    """List ALL participants (humans + agents) in the current team.

    Non-admin: any active team member may call this. Humans report
    ``online=False``, ``status="offline"``, ``last_seen=None`` (no presence by
    design); agents carry live presence exactly as ``GET /v1/agents`` does.
    """
    aweb_db = db.get_manager("aweb")

    rows = await aweb_db.fetch_all(
        """
        SELECT a.agent_id, a.alias, a.did_key, a.did_aw, a.address,
               a.human_name, a.agent_type, a.role
        FROM {{tables.agents}} a
        WHERE a.team_id = $1 AND a.deleted_at IS NULL
        ORDER BY a.alias
        """,
        identity.team_id,
    )

    # Presence from Redis (agents only; humans have no live presence).
    agent_ids = [str(r["agent_id"]) for r in rows]
    presences = (
        await list_agent_presences_by_workspace_ids(redis, agent_ids)
        if redis and agent_ids
        else []
    )
    presence_by_id = {str(p.get("workspace_id")): p for p in presences if p.get("workspace_id")}

    participants: list[ParticipantView] = []
    for r in rows:
        agent_id = str(r["agent_id"])
        kind = _participant_kind(r.get("agent_type"))
        alias = r.get("alias") or ""
        human_name = (r.get("human_name") or "").strip()
        display_name = human_name if (kind == "human" and human_name) else (human_name or alias)

        online = False
        status = "offline"
        last_seen = None
        role = r.get("role") or None
        # Humans never have presence; only join presence for agents.
        presence = presence_by_id.get(agent_id) if kind == "agent" else None
        if presence:
            online = True
            status = presence.get("status") or "active"
            last_seen = presence.get("last_seen") or None
            role = presence.get("role") or role

        participants.append(
            ParticipantView(
                kind=kind,
                alias=alias,
                display_name=display_name or alias,
                agent_id=agent_id,
                did_key=r.get("did_key") or None,
                did_aw=r.get("did_aw") or None,
                address=r.get("address") or None,
                role=role,
                agent_type=r.get("agent_type") or "agent",
                online=online,
                status=status,
                last_seen=last_seen,
            )
        )

    return ListParticipantsResponse(team_id=identity.team_id, participants=participants)
